docs/ibc/1_IBC_ARCHITECTURE.md

```diff
@@ -0,0 +1,151 @@
+| | +----------+     +-----+ | |        +---------+        | | +-----+     +----------+ | |
+| | | Module A | <-> | IBC | | | <----> | Relayer | <----> | | | IBC | <-> | Module B | | |
+| | +----------+     +-----+ | |        +---------+        | | +-----+     +----------+ | |
+| +--------------------------+ |                           | +--------------------------+ |
++------------------------------+                           +------------------------------+
+```
+
+## Operation
+
+The primary purpose of IBC is to provide reliable, authenticated, ordered communication between modules running on independent host distributed ledgers. This requires protocol logic in the following areas:
+- Data relay
+- Reliability
+- Flow control
+- Authentication
+- Statefulness
+- Multiplexing
+
+The following paragraphs outline the protocol logic within IBC for each area.
+
+### Data relay
+
+IBC assumes the existence of a set of relayer processes with access to an underlying network protocol stack (likely TCP/IP, UDP/IP, or QUIC/IP) and physical interconnect infrastructure. These relayer processes monitor a set of ledgers implementing the IBC protocol, continuously scanning the state of each ledger and executing transactions on another ledger when outgoing datagrams have been committed. For correct operation and progress in a connection between two ledgers, IBC requires only that at least one correct and live relayer process exists which can relay between the ledgers.
+
+### Reliability
+
+The network layer and relayer processes may behave in arbitrary ways, dropping, reordering, or duplicating packets, purposely attempting to send invalid transactions, or otherwise acting Byzantine. This must not compromise the safety or liveness of IBC. This is achieved by assigning a sequence number to each packet sent over an IBC connection, which is checked by the IBC handler on the receiving ledger, and providing a method for the sending ledger to check that the receiving ledger has in fact received and handled a packet before sending more packets or taking further action. Cryptographic commitments are used to prevent datagram forgery: the sending ledger commits to outgoing datagrams, and the receiving ledger checks these commitments, so datagrams altered in transit by a relayer will be rejected.
+
+### Flow control
+
+IBC does not require specific provision for computation-level flow control since the underlying ledgers will have throughput limitations and flow control mechanisms of their own (such as "gas" markets). Application-level flow control — limiting the rate of particular packets according to their content — may be useful to ensure security properties (limiting the value on a single ledger) and contain damage from Byzantine faults (allowing a challenge period to prove an equivocation, then closing a connection). IBC provides facilities for modules to reject packets and leaves particulars up to the higher-level application protocols.
```

Comment:
Pushing responsibility for logical security on application is the right way to build IBC but I feel we need a best practice guide soon after the start. Without a proper framework for reasoning about IBC application-level security IBC-relying applications won't be secure in practice.

Reply:
Definitely - by "application" I just mean that it isn't part of the "core" protocol - we'll want to specify and analyze concrete constructions for providing additional security in different ways.